### ISSUE TITLE
Configurable triggers based on terminal output

### ISSUE THREAD

**Issue by hktonylee** [wez/wezterm#1759]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I want to make my customized highlight for the terminal output. In iTerm, I can define regular expression to highlight date or Java namespace. And I can even send notification when the command outputs some specific output.

**Describe the solution you'd like**
Basically, this feature - https://iterm2.com/documentation-triggers.html

**Describe alternatives you've considered**
I think at least we need colorization, and run command.

**Additional context**
Here is some example as well - https://github.com/scottdware/iterm2-triggers

**Comment by kaddkaka** [comment]
Alternatives could be:

- Modify the command line tool output
- Use another command line tool that displays things more clearly
- Pipe the output through a tool that adds color to the output

Having a regexp check all text that is printed seems like heavy load? 🤔

**Comment by hktonylee** [comment]
Not everything can be pipe to other commands :( Let's say, I can't do that when checking log in production servers. Those servers do not allow installing color tool. And the log files can't leave the server too because of compliance / size problem.

If I had this feature, I could read the remote log with color -- This makes a huge difference.

Yes I agree running regexp is super heavy. But this is very convenient. I think users can decided whether to sacrifice the performance for this.

**Comment by digitallyserviced** [comment]
@hktonylee Just a heads up! You sure you cannot pipe to other commands, you should be able to pipe all the things! Just need to know the syntax. Have you tried to do the following?

```
ssh user@server.com 'tail -f /var/log/remoteserverlogfile.log' | /my/local/bin/logstalgia --opts 
# And the other way!
tail -f /var/log/mylocallogfile.log | ssh user@server.com '/usr/bin/remoteserverslogeater'
# even more fun the - (dash) when used after an option to specify a file usually makes it look at stdin
# unzip a tarball to a remote server (-C specifies destination)
tar zxf - somelocaltarball.tar.gz | ssh user@server.com 'tar zxf - -C /remoteserver/destination/'

```

Piping the data over the remote conn back to your local machine that has the colorizer or log viz or whatever tool that will consume a stdin pipe.  That may be more customizable and flexible then having to hack it into the Term.
You could write scripts in any language and have it output to (stdout) or consume (stdin) to do whatever you want with output from cmds.

**Comment by wez** [comment]
FWIW, if you're using `wezterm ssh` to connect to the remote network appliances mentioned in one of the links in the OP, then you can't simply pipe stuff.

I've been thinking about this request and looking at the possible actions in iterm2 and having a hard time seeing how to make those all fit nicely into wezterm, as there are some significant architectural differences and other cases that wezterm needs to support.

What I'm thinking here is to use the existing hyperlink rule stuff as a base to provide a way to manipulate the cell attributes for text that matches regular expressions.  Hyperlink rules allow setting an implied link destination, but we could also allow for setting any of the attributes (fg/bg color, bold, italic and so on).

The key difference from iTerm's trigger function is that this treatment would be done at render time, and does not rewrite text at tricky places in the terminal emulator.  Doing it at render time has the advantage of allowing you to live-reload your config to adjust the regex's to match the content.

I don't really want to support functions that transform the output beyond manipulating existing cell attributes at render time.

The reason is: while the terminal processes a stream of rendering instructions, those instructions are not guaranteed to be ordered in the same order as the text appears on the screen.  The application can eg: output `hello` starting with the `o` and generating a series of left cursor movements intermingled with the other letters until it has produced `hello` in the form that you see on the screen. TUI toolkits may produce crazily ordered output in order to optimize their display repainting operations, so any given run of text may be incomplete and a given line may be updated at any time.  Successfully transforming that kind of input stream is rather difficult.

A second reason is: with remote multiplexer connections, the output will already have been accumulated by the remote host and there isn't a place to insert stream processing/transformation.

If you want to perform actual transformation of the content, then you may wish to look at this custom event example which shows how to capture the output from a pane and open it in an editor in a new window: https://wezfurlong.org/wezterm/config/lua/wezterm/on.html#custom-events
That technique could be used to send the output through some custom processing/transformation instead of launching an editor.

**Comment by digitallyserviced** [comment]
@wez I agree, the only thing I think we would be missing after using custom events is the ability to create/render custom overlays using lua.

The overlays functionality you have for debug, search, copy, etc... would be a great way to tap into customizing interfaces or interaction with the terminal. 

Like maybe something that parses the pane text and keeps a running counter in an overlay that docks at the top or bottom. I know you could update stuff using the right-status but that is quite limited in size, as well as functionality. (doesn't seem to support hyperlinks or standard ansi color codes).

**Comment by wez** [comment]
> using the right-status but that is quite limited in size, as well as functionality. (doesn't seem to support hyperlinks or standard ansi color codes).

Yes it does!  [The example](https://wezfurlong.org/wezterm/config/lua/window/set_right_status.html) shows how to compose escapes using the [wezterm.format](https://wezfurlong.org/wezterm/config/lua/wezterm/format.html) function, but you can directly use ANSI escape sequences if you wish.

I haven't tried using hyperlinks in the right status area: it should know what they are, but the UI may not let you click them.

**Comment by digitallyserviced** [comment]
@wez I am quite aware of wezterm.format and "composing escapes" for coloring, but it is quite unwieldy and not DRY. The biggest thing that seems to be missing is the ability to pass a string literally and have it be parsed for osc7, csi, and other escapes. 

I should probably share my config since I have added quite a few things that I've tried to make wezterm a more flexible tool 

Mainly because the styling is a bit of a **PITA** having to define each attribute with the text, especially when a long string that may contain multiple groups of style attributes. 

- be able to do a 
```lua
local _, status, _ = wezterm.run_child_process({"cat" "~.cache/powerline-like-cmd-output"})
window.update_right_status(wezterm.format_non_inert(status))
```

- I know I could whip up a bunch of lua helpers to make the styling much easier that is required in my configs. The ability to define lua modules to be made available through the config too the global wezterm object would be nifty.

```lua
config.lua_modules.styles = require('plugins.styles')
wezterm.styles.add_attr_style("name", {{Attribute = {Foreground = {Color = "#fff"}})
table.insert(elements, {Style=wezterm.styles.name, Text = "ballz"}}
table.insert(elements, {Style=wezterm.styles.default, Text = " > "}}
wezterm.format(elements)
```

- wezterm.format doesn't seem to provide attributes relating to hyperlinks or clickable action type stuff like an attribute that provides a definable action that is emitted. 

I will be playing around and see what may be the most practical and maybe open enhancement issues if it would be something of interest...
